ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/RaceConditionIT.scala

```diff
@@ -0,0 +1,357 @@
+      )
+      ()
+    }
+  })
+
+  test(
+    "WWArchiveVsNonTransientCreate",
+    "Cannot re-create a contract if it hasn't yet been archived",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    for {
+      contract <- ledger.create(alice, ContractWithKey(alice))
+      _ <- Future.traverse(1 to Attempts) { attempt =>
+        if (attempt % 2 == 1) {
+          ledger.create(alice, ContractWithKey(alice)).transform(Success(_))
+        } else {
+          ledger.exercise(alice, contract.exerciseContractWithKey_Archive).transform(Success(_))
+        }
+      }
+      transactions <- ledger.transactionTrees(alice)
+    } yield {
+      import TransactionUtil._
+      def txLt(tx1: TransactionTree, tx2: TransactionTree): Boolean =
+        offsetLessThan(tx1.offset, tx2.offset)
+
+      val sorted = transactions.sortWith(txLt)
+
+      assert(
+        isCreateNonTransient(sorted.head),
+        "The first transaction is expected to be a contract creation",
+      )
+      assert(sorted.tail.nonEmpty, "Several transactions expected")
+
+      val (_, valid) = sorted.tail.foldLeft((sorted.head, true)) {
+        case ((previousTx, validUntilNow), currentTx) =>
+          if (validUntilNow) {
+            val valid = (isArchival(previousTx) && isCreateNonTransient(
+              currentTx
+            )) || (isCreateNonTransient(previousTx) && isArchival(currentTx))
+            (currentTx, valid)
+          } else {
+            (previousTx, validUntilNow)
+          }
+      }
+
+      if (!valid)
+        fail(s"""Invalid transaction sequence: ${sorted.map(printTransaction).mkString("\n")}""")
+    }
+  })
+
+  test(
+    "RWTransientCreateVsNonTransientCreate",
+    "Cannot create a transient contract and a non-transient contract with the same key",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    val ActionAt = 90
+    for {
+      wrapper <- ledger.create(alice, CreateWrapper(alice))
+      _ <- Future.traverse(1 to Attempts) { attempt =>
+        if (attempt == ActionAt) {
+          ledger.create(alice, ContractWithKey(alice)).map(_ => ()).transform(Success(_))
+        } else {
+          ledger
+            .exercise(alice, wrapper.exerciseCreateWrapper_CreateTransient)
+            .map(_ => ())
+            .transform(Success(_))
+        }
+      }
+      transactions <- ledger.transactionTrees(alice)
+    } yield {
+      import TransactionUtil._
+
+      val nonTransientCreateTransaction = transactions
+        .find(isCreateNonTransient)
+        .getOrElse(fail("No non-transient create transaction found"))
+
+      transactions
+        .filter(isTransientCreate)
+        .foreach(assertTransactionOrder(_, nonTransientCreateTransaction))
+    }
+  })
+
+  test(
+    "RWArchiveVsNonConsumingChoice",
+    "Cannot exercise a non-consuming choice after a contract archival",
```

Comment:
More generally, one cannot exercise any choice after a contract archival.

Reply:
I've updated the title.